### PR TITLE
[bct] Checker refactoring + new overload checker

### PR DESCRIPTION
**Pull request by catalinaperalta**
Adjusting the pluggable checker functionality to make it easier to migrate more checks. The design is still in flux but this will facilitate migration so that we can see how we need to refine/optimize the checker design.

Fixes https://github.com/Azure/azure-sdk-for-python/issues/36986

Tracking issue for future overload work for a different PR: https://github.com/Azure/azure-sdk-for-python/issues/37033

### CODE REVIEW

File: scripts/breaking_changes_checker/breaking_changes_tracker.py [hunk 1]
```diff
@@ -92,6 +92,7 @@ def __init__(self, stable: Dict, current: Dict, package_name: str, **kwargs: Any
         self.stable = stable
         self.current = current
         self.diff = jsondiff.diff(stable, current)
+        self.features_added = []
```
Comment (catalinaperalta) [inline scripts/breaking_changes_checker/breaking_changes_tracker.py:95]: Technically I think we will eventually move to having 1 changes list for all of the checkers specified on an instance of the tracker class but we can think more about what this will look like once the BreakingChangesTracker class is refactored to a more generic CodeReporter class.

File: scripts/breaking_changes_checker/_models.py [hunk 2]
```diff
@@ -23,6 +23,11 @@ class Suppression(NamedTuple):
     function_name: Optional[str] = None
     parameter_or_property_name: Optional[str] = None
 
+class CheckerType:
```
Comment (kristapratico) [inline scripts/breaking_changes_checker/_models.py:26]: Is this meant to be an enum?

File: scripts/breaking_changes_checker/checkers/added_method_overloads_checker.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python
+
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import sys
+import os
+sys.path.append(os.path.abspath("../../scripts/breaking_changes_checker"))
```
Comment (kristapratico) [inline scripts/breaking_changes_checker/checkers/added_method_overloads_checker.py:9]: why is this needed?

File: scripts/breaking_changes_checker/checkers/added_method_overloads_checker.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                            parsed_overload_signature = f"def {method_name}(" + ", ".join([f"{name}: {data['type']}" for name,  data in added_overload["parameters"].items()]) + ")"
+                            if added_overload["return_type"] is not None:
+                                parsed_overload_signature += f" -> {added_overload['return_type']}"
+                            changes_list.append((self.message["default"], self.name, module_name, class_name, method_name, parsed_overload_signature))
+                else:
+                    # this case is for when the overload is not a symbol and simply shows as a new overload in the diff
+                    parsed_overload_signature = f"def {method_name}(" + ", ".join([f"{name}: {data['type']}" for name,  data in overload["parameters"].items()]) + ")"
+                    if overload["return_type"] is not None:
+                        parsed_overload_signature += f" -> {overload['return_type']}"
+                    changes_list.append((self.message["default"], self.name, module_name, class_name, method_name, parsed_overload_signature))
```
Comment (kristapratico) [inline scripts/breaking_changes_checker/checkers/added_method_overloads_checker.py:42]: is this code the same as L33-36? could we share it?

**Review by kristapratico — APPROVED** [review]

File: scripts/breaking_changes_checker/checkers/added_method_overloads_checker.py (hunk 3 above)
Comment (catalinaperalta) [reply scripts/breaking_changes_checker/checkers/added_method_overloads_checker.py:9]: I wasnt able to reference the _models from here for some reason :/ I'll look back into why this was necessary in the future
